File: data/derived_cot/rq3_prompting/results/translation/qwen/py_to_cpp/code_output/UserLoginDB.cpp
#include <sqlite3.h>
#include <string>

class UserLoginDB {
private:
    sqlite3* connection;
    sqlite3_stmt* cursor;

public:
    explicit UserLoginDB(const std::string& db_name) {
        int rc = sqlite3_open(db_name.c_str(), &connection);
        if (rc != SQLITE_OK) {
            // Handle error (optional, matches Python's implicit behavior)
        }
        cursor = nullptr;
    }

    ~UserLoginDB() {
        if (cursor) {
            sqlite3_finalize(cursor);
        }
        sqlite3_close(connection);
    }

    bool insert_user(const std::string& username, const std::string& password) {
        const char* sql = "INSERT INTO users (username, password) VALUES (?, ?)";
        int rc = sqlite3_prepare_v2(connection, sql, -1, &cursor, nullptr);
        if (rc != SQLITE_OK) {
            return false;
        }

        sqlite3_bind_text(cursor, 1, username.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(cursor, 2, password.c_str(), -1, SQLITE_STATIC);

        int execute_rc = sqlite3_step(cursor);
        sqlite3_finalize(cursor);
        cursor = nullptr;

        if (execute_rc != SQLITE_DONE) {
            return false;
        }

        sqlite3_exec(connection, "COMMIT;", nullptr, nullptr, nullptr);
        return true;
    }

    std::vector<std::vector<std::string>> search_user_by_username(const std::string& username) {
        const char* sql = "SELECT * FROM users WHERE username = ?";
        int rc = sqlite3_prepare_v2(connection, sql, -1, &cursor, nullptr);
        if (rc != SQLITE_OK) {
            return {};
        }

        sqlite3_bind_text(cursor, 1, username.c_str(), -1, SQLITE_STATIC);
        int execute_rc = sqlite3_step(cursor);
        if (execute_rc != SQLITE_ROW && execute_rc != SQLITE_DONE) {
            sqlite3_finalize(cursor);
            cursor = nullptr;
            return {};
        }

        std::vector<std::vector<std::string>> result;
        while (execute_rc == SQLITE_ROW) {
            int num_columns = sqlite3_column_count(cursor);
            std::vector<std::string> row;
            for (int i = 0; i < num_columns; ++i) {
                const char* text = reinterpret_cast<const char*>(sqlite3_column_text(cursor, i));
                row.push_back(text ? std::string(text) : "");
            }
            result.push_back(row);
            execute_rc = sqlite3_step(cursor);
        }

        sqlite3_finalize(cursor);
        cursor = nullptr;
        return result;
    }

    bool delete_user_by_username(const std::string& username) {
        const char* sql = "DELETE FROM users WHERE username = ?";
        int rc = sqlite3_prepare_v2(connection, sql, -1, &cursor, nullptr);
        if (rc != SQLITE_OK) {
            return false;
        }

        sqlite3_bind_text(cursor, 1, username.c_str(), -1, SQLITE_STATIC);
        int execute_rc = sqlite3_step(cursor);
        sqlite3_finalize(cursor);
        cursor = nullptr;

        if (execute_rc != SQLITE_DONE) {
            return false;
        }

        sqlite3_exec(connection, "COMMIT;", nullptr, nullptr, nullptr);
        return true;
    }

    bool validate_user_login(const std::string& username, const std::string& password) {
        auto users = search_user_by_username(username);
        if (users.empty()) {
            return false;
        }

        auto& user = users[0];
        if (user.size() < 2 || user[1] != password) {
            return false;
        }

        return true;
    }
};

File: data/derived_cot/rq1_traces/translation/glm5.1/cpp_to_py/code_output/UserLoginDB.py
import sqlite3
from typing import Optional, Tuple

class UserLoginDB:
    def __init__(self, db_name: str):
        self.connection: Optional[sqlite3.Connection] = None
        
        try:
            # Connect to the SQLite database
            self.connection = sqlite3.connect(db_name)
        except sqlite3.Error as e:
            # Match C++ exception behavior
            raise RuntimeError(f"Cannot open database: {e}") from e
            
        self.create_table()

    def __del__(self):
        # Matches C++ destructor behavior
        self.close_connection()

    def insert_user(self, username: str, password: str) -> None:
        try:
            # Using parameterized queries to match sqlite3_bind_text behavior
            self.connection.execute(
                "INSERT INTO users (username, password) VALUES (?, ?)",
                (username, password)
            )
            # Commit to ensure data is saved immediately (matching C++ autocommit behavior)
            self.connection.commit()
        except sqlite3.Error as e:
            raise RuntimeError(f"Failed to insert user: {e}") from e

    def search_user_by_username(self, username: str) -> Optional[Tuple[str, str]]:
        try:
            cursor = self.connection.execute(
                "SELECT username, password FROM users WHERE username = ?",
                (username,)
            )
            row = cursor.fetchone()
            if row:
                # Returns a tuple of (username, password) if found, None otherwise
                # This perfectly matches the unique_ptr<tuple<string, string>> behavior
                return (row[0], row[1])
            return None
        except sqlite3.Error as e:
            raise RuntimeError(f"Failed to search user: {e}") from e

    def delete_user_by_username(self, username: str) -> None:
        try:
            self.connection.execute(
                "DELETE FROM users WHERE username = ?",
                (username,)
            )
            self.connection.commit()
        except sqlite3.Error as e:
            raise RuntimeError(f"Failed to delete user: {e}") from e

    def validate_user_login(self, username: str, password: str) -> bool:
        user = self.search_user_by_username(username)
        if user is not None and user[1] == password:
            return True
        return False

    def close_connection(self) -> None:
        if self.connection:
            self.connection.close()
            self.connection = None

    def create_table(self) -> None:
        create_table_query = """
            CREATE TABLE IF NOT EXISTS users (
                username TEXT,
                password TEXT
            )
        """
        try:
            self.connection.execute(create_table_query)
            self.connection.commit()
        except sqlite3.Error as e:
            raise RuntimeError(f"Cannot create table: {e}") from e